protocols/gossipsub/src/protocol.rs

```diff
@@ -0,0 +1,336 @@
+            .collect();
+
+        let prune_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_prune()
+            .into_iter()
+            .map(|mut prune| GossipsubControlAction::Prune {
+                topic_hash: TopicHash::from_raw(prune.take_topicID()),
+            })
+            .collect();
+
+        control_msgs.extend(ihave_msgs);
+        control_msgs.extend(iwant_msgs);
+        control_msgs.extend(graft_msgs);
+        control_msgs.extend(prune_msgs);
+
+        Ok(Some(GossipsubRpc {
+            messages,
+            subscriptions: rpc
+                .take_subscriptions()
+                .into_iter()
+                .map(|mut sub| GossipsubSubscription {
+                    action: if sub.get_subscribe() {
+                        GossipsubSubscriptionAction::Subscribe
+                    } else {
+                        GossipsubSubscriptionAction::Unsubscribe
+                    },
+                    topic_hash: TopicHash::from_raw(sub.take_topicid()),
+                })
+                .collect(),
+            control_msgs,
+        }))
+    }
+}
+
+/// A message received by the gossipsub system.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct GossipsubMessage {
+    /// Id of the peer that published this message.
+    pub source: PeerId,
+
+    /// Content of the message. Its meaning is out of scope of this library.
+    pub data: Vec<u8>,
+
+    /// A random sequence number.
+    pub sequence_number: u64,
+
+    /// List of topics this message belongs to.
+    ///
+    /// Each message can belong to multiple topics at once.
+    pub topics: Vec<TopicHash>,
+}
+
+impl GossipsubMessage {
+    /// Converts message into a message_id.
+    // To be compatible with the go implementation
+    pub fn id(&self) -> String {
+        let mut source_string = self.source.to_base58();
+        source_string.push_str(&self.sequence_number.to_string());
+        source_string
+    }
+}
+
+/// A subscription received by the gossipsub system.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct GossipsubSubscription {
+    /// Action to perform.
+    pub action: GossipsubSubscriptionAction,
+    /// The topic from which to subscribe or unsubscribe.
+    pub topic_hash: TopicHash,
+}
+
+/// Action that a subscription wants to perform.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum GossipsubSubscriptionAction {
+    /// The remote wants to subscribe to the given topic.
+    Subscribe,
+    /// The remote wants to unsubscribe from the given topic.
+    Unsubscribe,
+}
+
+/// A Control message received by the gossipsub system.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum GossipsubControlAction {
+    /// Node broadcasts known messages per topic - IHave control message.
+    IHave {
+        /// The topic of the messages.
+        topic_hash: TopicHash,
+        /// A list of known message ids (peer_id + sequence _number) as a string.
+        message_ids: Vec<String>,
```

Comment:
I would suggest leveraging the fact that we have a type system at hand and use the [new type pattern](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html) to define message ids. Makes it more descriptive and less error prone (checked at compile time). What do you think?